Node 22.5.0 started to crash and hangs on different cases

### Version

22.5.0

### Platform

```text
macOS 15.0, M1
```

### What steps will reproduce the bug?

Now I haven't much time to properly debug it, so a couple of cases:

Clone [`core-js` repo](https://github.com/zloirock/core-js).

`npm i` just hangs on.

Ok, if we somehow install dependencies and run `npm test`, Node crashes with:
```
FATAL ERROR: v8::Object::GetCreationContextChecked No creation context available
----- Native stack trace -----

 1: 0x104a4e230 node::OnFatalError(char const*, char const*) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
 2: 0x104c20898 v8::Utils::ReportApiFailure(char const*, char const*) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
 3: 0x104c358fc v8::Object::GetCreationContextChecked() [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
 4: 0x104a55f88 node::fs::FastClose(v8::Local<v8::Object>, int, v8::FastApiCallbackOptions&) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
 5: 0x10b9accc4 
 6: 0x1056a6c0c Builtins_JSEntryTrampoline [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
 7: 0x1056a68f4 Builtins_JSEntry [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
 8: 0x104d50564 v8::internal::(anonymous namespace)::Invoke(v8::internal::Isolate*, v8::internal::(anonymous namespace)::InvokeParams const&) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
 9: 0x104d4fadc v8::internal::Execution::Call(v8::internal::Isolate*, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>, int, v8::internal::Handle<v8::internal::Object>*) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
10: 0x104c369cc v8::Function::Call(v8::Local<v8::Context>, v8::Local<v8::Value>, int, v8::Local<v8::Value>*) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
11: 0x104970fa0 node::InternalMakeCallback(node::Environment*, v8::Local<v8::Object>, v8::Local<v8::Object>, v8::Local<v8::Function>, int, v8::Local<v8::Value>*, node::async_context) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
12: 0x104987d8c node::AsyncWrap::MakeCallback(v8::Local<v8::Function>, int, v8::Local<v8::Value>*) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
13: 0x104a53ca4 node::fs::FSReqCallback::Resolve(v8::Local<v8::Value>) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
14: 0x104a549a8 node::fs::AfterInteger(uv_fs_s*) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
15: 0x104a482fc node::MakeLibuvRequestCallback<uv_fs_s, void (*)(uv_fs_s*)>::Wrapper(uv_fs_s*) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
16: 0x105684a70 uv__work_done [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
17: 0x105688544 uv__async_io [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
18: 0x10569ab08 uv__io_poll [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
19: 0x105688b08 uv_run [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
20: 0x104971714 node::SpinEventLoopInternal(node::Environment*) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
21: 0x104a8f8f0 node::NodeMainInstance::Run(node::ExitCode*, node::Environment*) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
22: 0x104a8f688 node::NodeMainInstance::Run() [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
23: 0x104a15928 node::Start(int, char**) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
24: 0x18b2cf274 start [/usr/lib/dyld]


#
# Fatal error in , line 0
# Missing deoptimization information for OptimizedFrame::Summarize.
#
#
#
#FailureMessage Object: 0x16b488648
----- Native stack trace -----

 1: 0x104ab7a4c node::NodePlatform::GetStackTracePrinter()::$_3::__invoke() [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
 2: 0x105f3e5d4 V8_Fatal(char const*, ...) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
 3: 0x104d5993c v8::internal::OptimizedFrame::Summarize(std::__1::vector<v8::internal::FrameSummary, std::__1::allocator<v8::internal::FrameSummary>>*) const [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
 4: 0x104d62c60 v8::internal::Isolate::CaptureDetailedStackTrace(int, v8::StackTrace::StackTraceOptions) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
 5: 0x104c2ace0 v8::StackTrace::CurrentStackTrace(v8::Isolate*, int, v8::StackTrace::StackTraceOptions) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
 6: 0x104a4cc0c node::GetCurrentStackTrace(v8::Isolate*, int) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
 7: 0x1049b869c node::DumpJavaScriptBacktrace(__sFILE*) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
 8: 0x104a4e238 node::OnFatalError(char const*, char const*) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
 9: 0x104c20898 v8::Utils::ReportApiFailure(char const*, char const*) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
10: 0x104c358fc v8::Object::GetCreationContextChecked() [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
11: 0x104a55f88 node::fs::FastClose(v8::Local<v8::Object>, int, v8::FastApiCallbackOptions&) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
12: 0x10b9accc4 
13: 0x1056a6c0c Builtins_JSEntryTrampoline [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
14: 0x1056a68f4 Builtins_JSEntry [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
15: 0x104d50564 v8::internal::(anonymous namespace)::Invoke(v8::internal::Isolate*, v8::internal::(anonymous namespace)::InvokeParams const&) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
16: 0x104d4fadc v8::internal::Execution::Call(v8::internal::Isolate*, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>, int, v8::internal::Handle<v8::internal::Object>*) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
17: 0x104c369cc v8::Function::Call(v8::Local<v8::Context>, v8::Local<v8::Value>, int, v8::Local<v8::Value>*) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
18: 0x104970fa0 node::InternalMakeCallback(node::Environment*, v8::Local<v8::Object>, v8::Local<v8::Object>, v8::Local<v8::Function>, int, v8::Local<v8::Value>*, node::async_context) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
19: 0x104987d8c node::AsyncWrap::MakeCallback(v8::Local<v8::Function>, int, v8::Local<v8::Value>*) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
20: 0x104a53ca4 node::fs::FSReqCallback::Resolve(v8::Local<v8::Value>) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
21: 0x104a549a8 node::fs::AfterInteger(uv_fs_s*) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
22: 0x104a482fc node::MakeLibuvRequestCallback<uv_fs_s, void (*)(uv_fs_s*)>::Wrapper(uv_fs_s*) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
23: 0x105684a70 uv__work_done [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
24: 0x105688544 uv__async_io [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
25: 0x10569ab08 uv__io_poll [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
26: 0x105688b08 uv_run [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
27: 0x104971714 node::SpinEventLoopInternal(node::Environment*) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
28: 0x104a8f8f0 node::NodeMainInstance::Run(node::ExitCode*, node::Environment*) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
29: 0x104a8f688 node::NodeMainInstance::Run() [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
30: 0x104a15928 node::Start(int, char**) [/Users/denispuskarev/.nvm/versions/node/v22.5.0/bin/node]
31: 0x18b2cf274 start [/usr/lib/dyld]
/bin/bash: line 1: 92352 Trace/BPT trap: 5       npm run build-compat
```
on [this script](https://github.com/zloirock/core-js/blob/master/scripts/build-compat/entries.mjs).

The same error happens, for example, if we somehow built it (for example, with the previous Node version) and run other scripts like [this](https://github.com/zloirock/core-js/blob/master/scripts/check-unused-modules.mjs).

In most of those cases used [`zx`](https://www.npmjs.com/package/zx) and / or [`konan`](https://www.npmjs.com/package/konan).

Sure, before Node 22.5.0, everything worked fine.

we are seeing this in bunch of builds  in this build yarn cannot find the prettier this worked the nodejs version 22.4.1 

```
 time=2024-07-17T20:45:51.408Z level=INFO msg="[git checkout] repo='https://github.com/replicatedhq/kots' dest='.' depth='1' branch='' tag='v1.112.1' expcommit='07dc36e669e3c33f6a70cb8d1efd1539a0b6a386' recurse='false'" pkg=kots
  time=2024-07-17T20:45:51.414Z level=INFO msg="[git checkout] execute: git config --global --add safe.directory /tmp/tmp.m9Ey9U" pkg=kots
  time=2024-07-17T20:45:51.416Z level=INFO msg="[git checkout] execute: git config --global --add safe.directory /home/build" pkg=kots
  time=2024-07-17T20:45:51.418Z level=INFO msg="[git checkout] execute: git clone --quiet --origin=origin --config=user.name=Melange Build --config=user.email=melange-build@cgr.dev --config=advice.detachedHead=false --branch=v1.112.1 --depth=1 https://github.com/replicatedhq/kots /tmp/tmp.m9Ey9U" pkg=kots
  time=2024-07-17T20:45:52.837Z level=INFO msg="[git checkout] execute: cd /tmp/tmp.m9Ey9U" pkg=kots
  time=2024-07-17T20:45:52.837Z level=INFO msg="[git checkout] tar -c . | tar -C \"/home/build\" -x" pkg=kots
  time=2024-07-17T20:45:53.542Z level=INFO msg="[git checkout] execute: cd /home/build" pkg=kots
  time=2024-07-17T20:45:53.542Z level=INFO msg="[git checkout] execute: git config --global --add safe.directory /home/build" pkg=kots
  time=2024-07-17T20:45:53.545Z level=INFO msg="[git checkout] execute: git fetch --quiet origin --depth=1 --no-tags +refs/tags/v1.112.1:refs/origin/tags/v1.112.1" pkg=kots
  time=2024-07-17T20:45:53.788Z level=INFO msg="[git checkout] execute: git checkout --quiet origin/tags/v1.112.1" pkg=kots
  time=2024-07-17T20:45:53.928Z level=INFO msg="[git checkout] tag v1.112.1 is 07dc36e669e3c33f6a70cb8d1efd1539a0b6a386" pkg=kots
  time=2024-07-17T20:45:53.938Z level=WARN msg="+ mkdir -p /home/build/melange-out/kots/etc" pkg=kots
  time=2024-07-17T20:45:53.939Z level=WARN msg="+ mkdir -p /home/build/melange-out/kots/usr/bin" pkg=kots
  time=2024-07-17T20:45:53.941Z level=WARN msg="+ mv deploy/assets/backup.sh deploy/assets/fs-minio-check.sh deploy/assets/fs-minio-keys-sha.sh deploy/assets/fs-minio-reset.sh deploy/assets/kots-upgrade.sh deploy/assets/migrate-s3.sh deploy/assets/postgres deploy/assets/restore-db.sh deploy/assets/restore-s3.sh deploy/assets/restore.sh deploy/assets/s3-bucket-create.sh deploy/assets/s3-bucket-head.sh /home/build/melange-out/kots/etc/" pkg=kots
  time=2024-07-17T20:45:53.943Z level=WARN msg="+ export 'VERSION=1.112.1'" pkg=kots
  time=2024-07-17T20:45:53.943Z level=WARN msg="+ export 'GIT_TAG=1.112.1'" pkg=kots
  time=2024-07-17T20:45:53.943Z level=WARN msg="+ source .image.env" pkg=kots
  time=2024-07-17T20:45:53.943Z level=WARN msg="+ MINIO_TAG=0.20240713.014615-r0" pkg=kots
  time=2024-07-17T20:45:53.943Z level=WARN msg="+ RQLITE_TAG=8.26.7-r0" pkg=kots
  time=2024-07-17T20:45:53.943Z level=WARN msg="+ DEX_TAG=2.40.0-r3" pkg=kots
  time=2024-07-17T20:45:53.943Z level=WARN msg="+ SCHEMAHERO_TAG=0.17.9" pkg=kots
  time=2024-07-17T20:45:53.943Z level=WARN msg="+ LVP_TAG=v0.6.7" pkg=kots
  time=2024-07-17T20:45:53.943Z level=WARN msg="+ KOTS_KUSTOMIZE_BIN_DIR=/usr/bin" pkg=kots
  time=2024-07-17T20:45:53.943Z level=WARN msg="+ export 'PACT_SKIP_BINARY_INSTALL=true'" pkg=kots
  time=2024-07-17T20:45:53.943Z level=WARN msg="+ yarn install --pure-lockfile --network-concurrency 1" pkg=kots
  time=2024-07-17T20:45:54.190Z level=INFO msg="yarn install v1.22.22" pkg=kots
  time=2024-07-17T20:45:54.203Z level=INFO msg="info No lockfile found." pkg=kots
  time=2024-07-17T20:45:54.207Z level=INFO msg="[1/4] Resolving packages..." pkg=kots
  time=2024-07-17T20:45:54.207Z level=INFO msg="[2/4] Fetching packages..." pkg=kots
  time=2024-07-17T20:45:54.208Z level=INFO msg="[3/4] Linking dependencies..." pkg=kots
  time=2024-07-17T20:45:54.212Z level=INFO msg="[4/4] Building fresh packages..." pkg=kots
  time=2024-07-17T20:45:54.216Z level=INFO msg="Done in 0.03s." pkg=kots
  time=2024-07-17T20:45:54.235Z level=WARN msg="+ make -C web deps lint build-kotsadm" pkg=kots
  time=2024-07-17T20:45:54.237Z level=INFO msg="make: Entering directory '/home/build/web'" pkg=kots
  time=2024-07-17T20:45:54.244Z level=INFO msg="yarn --silent --frozen-lockfile" pkg=kots
  time=2024-07-17T20:45:56.521Z level=INFO msg="yarn prettier && yarn lint && yarn test:typecheck" pkg=kots
  time=2024-07-17T20:45:56.723Z level=INFO msg="yarn run v1.22.22" pkg=kots
  time=2024-07-17T20:45:56.763Z level=INFO msg="$ prettier src --check" pkg=kots
  time=2024-07-17T20:45:56.775Z level=WARN msg="/bin/sh: prettier: not found" pkg=kots
  time=2024-07-17T20:45:56.785Z level=WARN msg="error Command failed with exit code 127." pkg=kots
  time=2024-07-17T20:45:56.785Z level=INFO msg="info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command." pkg=kots
  time=2024-07-17T20:45:56.810Z level=INFO msg="make: Leaving directory '/home/build/web'" pkg=kots
  time=2024-07-17T20:45:56.810Z level=WARN msg="make: *** [Makefile:55: lint] Error 127" pkg=kots
  time=2024-07-17T20:45:56.811Z level=ERROR msg="building package: unable to run package kots pipeline: unable to run pipeline: exit status 2" pkg=kots
  ```

--- a/src/env.cc
+++ b/src/env.cc
@@ -1841,23 +1841,12 @@ void Environment::AddUnmanagedFd(int fd) {
   }
 }
 
-void Environment::RemoveUnmanagedFd(int fd, bool schedule_native_immediate) {
+void Environment::RemoveUnmanagedFd(int fd) {
   if (!tracks_unmanaged_fds()) return;
   size_t removed_count = unmanaged_fds_.erase(fd);
   if (removed_count == 0) {
-    if (schedule_native_immediate) {
-      SetImmediateThreadsafe([&](Environment* env) {
-        ProcessEmitWarning(this,
-                           "File descriptor %d closed but not opened in "
-                           "unmanaged mode",
-                           fd);
-      });
-    } else {
-      ProcessEmitWarning(
-          this,
-          "File descriptor %d closed but not opened in unmanaged mode",
-          fd);
-    }
+    ProcessEmitWarning(
+        this, "File descriptor %d closed but not opened in unmanaged mode", fd);
   }
 }
 

--- a/src/env.h
+++ b/src/env.h
@@ -1027,7 +1027,7 @@ class Environment : public MemoryRetainer {
   std::unique_ptr<v8::BackingStore> release_managed_buffer(const uv_buf_t& buf);
 
   void AddUnmanagedFd(int fd);
-  void RemoveUnmanagedFd(int fd, bool schedule_native_immediate = false);
+  void RemoveUnmanagedFd(int fd);
 
   template <typename T>
   void ForEachRealm(T&& iterator) const;

--- a/src/node_external_reference.h
+++ b/src/node_external_reference.h
@@ -47,9 +47,6 @@ using CFunctionCallbackWithUint8ArrayUint32Int64Bool =
                 uint32_t,
                 int64_t,
                 bool);
-using CFunctionWithObjectInt32Fallback = void (*)(v8::Local<v8::Object>,
-                                                  int32_t input,
-                                                  v8::FastApiCallbackOptions&);
 using CFunctionWithUint32 = uint32_t (*)(v8::Local<v8::Value>,
                                          const uint32_t input);
 using CFunctionWithDoubleReturnDouble = double (*)(v8::Local<v8::Value>,
@@ -78,7 +75,6 @@ class ExternalReferenceRegistry {
   V(CFunctionCallbackWithTwoUint8Arrays)                                       \
   V(CFunctionCallbackWithTwoUint8ArraysFallback)                               \
   V(CFunctionCallbackWithUint8ArrayUint32Int64Bool)                            \
-  V(CFunctionWithObjectInt32Fallback)                                          \
   V(CFunctionWithUint32)                                                       \
   V(CFunctionWithDoubleReturnDouble)                                           \
   V(CFunctionWithInt64Fallback)                                                \

--- a/src/node_file.cc
+++ b/src/node_file.cc
@@ -54,7 +54,6 @@ namespace fs {
 
 using v8::Array;
 using v8::BigInt;
-using v8::CFunction;
 using v8::Context;
 using v8::EscapableHandleScope;
 using v8::FastApiCallbackOptions;
@@ -299,7 +298,7 @@ FileHandle::TransferData::~TransferData() {
 
 BaseObjectPtr<BaseObject> FileHandle::TransferData::Deserialize(
     Environment* env,
-    Local<v8::Context> context,
+    v8::Local<v8::Context> context,
     std::unique_ptr<worker::TransferData> self) {
   BindingData* bd = Realm::GetBindingData<BindingData>(context);
   if (bd == nullptr) return {};
@@ -961,7 +960,7 @@ void Access(const FunctionCallbackInfo<Value>& args) {
   }
 }
 
-static void Close(const FunctionCallbackInfo<Value>& args) {
+void Close(const FunctionCallbackInfo<Value>& args) {
   Environment* env = Environment::GetCurrent(args);
 
   const int argc = args.Length();
@@ -986,30 +985,6 @@ static void Close(const FunctionCallbackInfo<Value>& args) {
     FS_SYNC_TRACE_END(close);
   }
 }
-
-static void FastClose(Local<Object> recv,
-                      int32_t fd,
-                      // NOLINTNEXTLINE(runtime/references) This is V8 api.
-                      v8::FastApiCallbackOptions& options) {
-  Environment* env = Environment::GetCurrent(recv->GetCreationContextChecked());
-
-  uv_fs_t req;
-  FS_SYNC_TRACE_BEGIN(close);
-  int err = uv_fs_close(nullptr, &req, fd, nullptr) < 0;
-  FS_SYNC_TRACE_END(close);
-  uv_fs_req_cleanup(&req);
-
-  if (err < 0) {
-    options.fallback = true;
-  } else {
-    // Note: Only remove unmanaged fds if the close was successful.
-    // RemoveUnmanagedFd() can call ProcessEmitWarning() which calls back into
-    // JS process.emitWarning() and violates the fast API protocol.
-    env->RemoveUnmanagedFd(fd, true /* schedule native immediate */);
-  }
-}
-
-CFunction fast_close_ = CFunction::Make(FastClose);
 
 static void ExistsSync(const FunctionCallbackInfo<Value>& args) {
   Environment* env = Environment::GetCurrent(args);
@@ -3330,7 +3305,7 @@ static void CreatePerIsolateProperties(IsolateData* isolate_data,
             "getFormatOfExtensionlessFile",
             GetFormatOfExtensionlessFile);
   SetMethod(isolate, target, "access", Access);
-  SetFastMethod(isolate, target, "close", Close, &fast_close_);
+  SetMethod(isolate, target, "close", Close);
   SetMethod(isolate, target, "existsSync", ExistsSync);
   SetMethod(isolate, target, "open", Open);
   SetMethod(isolate, target, "openFileHandle", OpenFileHandle);
@@ -3455,8 +3430,6 @@ void RegisterExternalReferences(ExternalReferenceRegistry* registry) {
 
   registry->Register(GetFormatOfExtensionlessFile);
   registry->Register(Close);
-  registry->Register(FastClose);
-  registry->Register(fast_close_.GetTypeInfo());
   registry->Register(ExistsSync);
   registry->Register(Open);
   registry->Register(OpenFileHandle);

